### PR TITLE
ui: Native macOS implementation for setting cursor shape

### PR DESCRIPTION
**Pull request by tomlister**
- Adds CursorShapeHorizontalResize & CursorShapeVerticalResize.

**Need someone to test on Windows and Linux.**

re: #1571

### CODE REVIEW

File: internal/uidriver/glfw/ui_darwin.go [hunk 1]
```diff
@@ -43,9 +43,35 @@ package glfw
 //   return [[NSApplication sharedApplication] currentSystemPresentationOptions] &
 //       NSApplicationPresentationFullScreen;
 // }
+//
+// static void setNativeCursor(int cursorID) {
+//   id cursor = [[NSCursor class] performSelector:@selector(arrowCursor)];
+//   switch(cursorID){
+//     case 0:
```
Comment (hajimehoshi) [inline internal/uidriver/glfw/ui_darwin.go:50]: I prefer unindent case. So the code will be like this:

```c
switch(cursorID){
case 0:
  cursor = [[NSCursor class] performSelector:@selector(arrowCursor)];
  break
```

File: internal/uidriver/glfw/ui_unix.go [hunk 2]
```diff
@@ -67,3 +68,8 @@ func (u *UserInterface) nativeWindow() uintptr {
 func (u *UserInterface) isNativeFullscreen() bool {
 	return false
 }
+
+func (u *UserInterface) setNativeCursor(shape driver.CursorShape) {
+	// TODO: Use native API in the future (issue #1571)
```
Comment (hajimehoshi) [inline internal/uidriver/glfw/ui_unix.go:73]: s/issue #1571/#1571/

File: internal/driver/cursor.go [hunk 3]
```diff
@@ -29,4 +29,6 @@ const (
 	CursorShapeText
 	CursorShapeCrosshair
 	CursorShapePointer
+	CursorShapeHorizontalResize
+	CursorShapeVerticalResize
```
Comment (hajimehoshi) [inline internal/driver/cursor.go:33]: Use CSS's cursor names: https://developer.mozilla.org/en-US/docs/Web/CSS/cursor

```
CursorShapeEWResize
CursorShapeNSResize
```

File: internal/uidriver/glfw/ui_unix.go (hunk 2 above)
Comment (tomlister) [reply internal/uidriver/glfw/ui_unix.go:73]: So just replace (issue #1571) with that?

Comment (hajimehoshi) [reply internal/uidriver/glfw/ui_unix.go:73]: With `#1571`

Comment (hajimehoshi) [reply internal/uidriver/glfw/ui_unix.go:73]: I meant to eliminate the prefix `issue `.

File: internal/uidriver/glfw/ui_unix.go [hunk 4]
```diff
@@ -67,3 +68,8 @@ func (u *UserInterface) nativeWindow() uintptr {
 func (u *UserInterface) isNativeFullscreen() bool {
 	return false
 }
+
+func (u *UserInterface) setNativeCursor(shape driver.CursorShape) {
+	// TODO: Use native API in the future #1571
```
Comment (hajimehoshi) [inline internal/uidriver/glfw/ui_unix.go:73]: Please keeps the parenthesis.

`(#1571)`

File: internal/uidriver/glfw/ui_darwin.go [hunk 5]
```diff
@@ -43,9 +43,35 @@ package glfw
 //   return [[NSApplication sharedApplication] currentSystemPresentationOptions] &
 //       NSApplicationPresentationFullScreen;
 // }
+//
+// static void setNativeCursor(int cursorID) {
+//   id cursor = [[NSCursor class] performSelector:@selector(arrowCursor)];
+//   switch(cursorID){
```
Comment (hajimehoshi) [inline internal/uidriver/glfw/ui_darwin.go:49]: Please insert spaces like `switch (cursorID) {`

**Review by hajimehoshi — COMMENTED** [review]
Thank you, this almost looks good!

**Comment by hajimehoshi** [comment]
I'll try this change on my Windows and Linux machine later.

**Comment by hajimehoshi** [comment]
Let's update examples/cursor to use the new shapes.

**Comment by tomlister** [comment]
Updated the example.

**Comment by tomlister** [comment]
That should not be failing...

File: examples/cursor/main.go [hunk 6]
```diff
@@ -71,22 +75,21 @@ func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
...
-			image.Rect(100, 100, 200, 300): ebiten.CursorShapeDefault,
-			image.Rect(200, 100, 300, 300): ebiten.CursorShapeText,
-			image.Rect(300, 100, 400, 300): ebiten.CursorShapeCrosshair,
-			image.Rect(400, 100, 500, 300): ebiten.CursorShapePointer,
+			image.Rect(70, 100, 170, 300):  ebiten.CursorShapeDefault,
+			image.Rect(170, 100, 270, 300): ebiten.CursorShapeText,
+			image.Rect(270, 100, 370, 300): ebiten.CursorShapeCrosshair,
+			image.Rect(370, 100, 470, 200): ebiten.CursorShapeEWResize,
+			image.Rect(370, 200, 470, 300): ebiten.CursorShapeNSResize,
+			image.Rect(470, 100, 570, 300): ebiten.CursorShapePointer,
```
Comment (hajimehoshi) [inline examples/cursor/main.go:83]: Hmm, let's make this more consistent like:

```
			image.Rect(100, 100, 200, 200):  ebiten.CursorShapeDefault,
			image.Rect(200, 100, 300, 200): ebiten.CursorShapeText,
			image.Rect(300, 100, 400, 200): ebiten.CursorShapeCrosshair,
			image.Rect(100, 200, 200, 300): ebiten.CursorShapeEWResize,
			image.Rect(200, 200, 300, 300): ebiten.CursorShapeNSResize,
			image.Rect(300, 200, 400, 300): ebiten.CursorShapePointer,
```

Comment (tomlister) [reply examples/cursor/main.go:83]: Will get onto this soon.

File: examples/cursor/main.go [hunk 7]
```diff
@@ -71,22 +75,21 @@ func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
...
+			image.Rect(200, 200, 300, 300): ebiten.CursorShapeNSResize,
+			image.Rect(300, 200, 400, 300): ebiten.CursorShapePointer,
 		},
 		gridColors: map[image.Rectangle]color.Color{},
 	}
 	for rect, c := range g.grids {
 		clr := color.RGBA{0x40, 0x40, 0x40, 0xff}
-		switch c % 3 {
-		case 0:
+		if c == 0 {
```
Comment (hajimehoshi) [inline examples/cursor/main.go:89]: This should be `c % 2 == 0` to make each tile distinguishable.

File: examples/cursor/main.go [hunk 8]
```diff
@@ -71,22 +75,21 @@ func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
...
-			image.Rect(100, 100, 200, 300): ebiten.CursorShapeDefault,
-			image.Rect(200, 100, 300, 300): ebiten.CursorShapeText,
-			image.Rect(300, 100, 400, 300): ebiten.CursorShapeCrosshair,
-			image.Rect(400, 100, 500, 300): ebiten.CursorShapePointer,
+			image.Rect(100, 100, 200, 200): ebiten.CursorShapeDefault,
+			image.Rect(200, 100, 300, 200): ebiten.CursorShapeText,
+			image.Rect(300, 100, 400, 200): ebiten.CursorShapeCrosshair,
+			image.Rect(100, 200, 200, 300): ebiten.CursorShapeEWResize,
+			image.Rect(200, 200, 300, 300): ebiten.CursorShapeNSResize,
+			image.Rect(300, 200, 400, 300): ebiten.CursorShapePointer,
```
Comment (hajimehoshi) [inline examples/cursor/main.go:83]: Sorry, let's keep the cursor order. Then

```
			image.Rect(100, 100, 200, 200): ebiten.CursorShapeDefault,
			image.Rect(200, 100, 300, 200): ebiten.CursorShapeText,
			image.Rect(300, 100, 400, 200): ebiten.CursorShapeCrosshair,
			image.Rect(100, 200, 200, 300): ebiten.CursorShapePointer,
			image.Rect(200, 200, 300, 300): ebiten.CursorShapeEWResize,
			image.Rect(300, 200, 400, 300): ebiten.CursorShapeNSResize,
```

File: examples/cursor/main.go [hunk 9]
```diff
@@ -59,6 +59,10 @@ func (g *Game) Draw(screen *ebiten.Image) {
 		ebitenutil.DebugPrint(screen, "CursorShape: Text")
 	case ebiten.CursorShapeCrosshair:
 		ebitenutil.DebugPrint(screen, "CursorShape: Crosshair")
+	case ebiten.CursorShapeEWResize:
+		ebitenutil.DebugPrint(screen, "CursorShape: EW Resize")
+	case ebiten.CursorShapeNSResize:
+		ebitenutil.DebugPrint(screen, "CursorShape: NS Resize")
 	case ebiten.CursorShapePointer:
 		ebitenutil.DebugPrint(screen, "CursorShape: Pointer")
```
Comment (hajimehoshi) [inline examples/cursor/main.go:67]: Let's keep the cursor definition order.

```
	case ebiten.CursorShapePointer:
		ebitenutil.DebugPrint(screen, "CursorShape: Pointer")
	case ebiten.CursorShapeEWResize:
		ebitenutil.DebugPrint(screen, "CursorShape: EW Resize")
	case ebiten.CursorShapeNSResize:
		ebitenutil.DebugPrint(screen, "CursorShape: NS Resize")
```

**Comment by tomlister** [comment]
I'll get to these later tonight

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!

**Pull request merged**